go/ql/test/experimental/CWE-522-DecompressionBombs/test.go

```diff
@@ -0,0 +1,655 @@
+
+	zstdReader, _ := zstdKlauspost.NewReader(file)
+	var out []byte = make([]byte, 70)
+	zstdReader.Read(out) // $ hasValueFlow="zstdReader"
+	var buf bytes.Buffer
+	zstdReader.WriteTo(&buf) // $ hasValueFlow="zstdReader"
+	var src []byte
+	zstdReader.DecodeAll(src, nil) // $ hasValueFlow="zstdReader"
+	tarRead = tar.NewReader(zstdReader)
+
+	TarDecompressor(tarRead)
+}
+func Zstd_KlauspostSafe(file io.Reader) {
+	zstdReader, _ := zstdKlauspost.NewReader(file)
+	var out []byte = make([]byte, 70)
+	var totalBytes int64 = 0
+	i := 1
+	for i > 0 {
+		i, _ = zstdReader.Read(out) // GOOD: The output size is being controlled
+		fmt.Println("do sth with output:", zstdReader)
+		totalBytes = totalBytes + int64(i)
+		if totalBytes > 1024*1024*1024*5 {
+			fmt.Print(totalBytes, "exceeded")
+			break
+		}
+	}
+}
+func Zstd_DataDog(file io.Reader) {
+	var tarRead *tar.Reader
+
+	zstdReader := zstdDataDog.NewReader(file)
+	var out []byte = make([]byte, 70)
+	zstdReader.Read(out) // $ hasValueFlow="zstdReader"
+	tarRead = tar.NewReader(zstdReader)
+
+	TarDecompressor(tarRead)
+}
+func Zstd_DataDogSafe(file io.Reader) {
+	zstdReader := zstdDataDog.NewReader(file)
+	var out []byte = make([]byte, 70)
+	var totalBytes int64 = 0
+	i := 1
+	for i > 0 {
+		i, _ = zstdReader.Read(out) // GOOD: The output size is being controlled
+		fmt.Println("do sth with output:", zstdReader)
+		totalBytes = totalBytes + int64(i)
+		if totalBytes > 1024*1024*1024*5 {
+			fmt.Print(totalBytes, "exceeded")
+			break
+		}
+	}
+}
+func Xz(file io.Reader) {
+	var tarRead *tar.Reader
+
+	xzReader, _ := xz.NewReader(file)
+	var out []byte = make([]byte, 70)
+	xzReader.Read(out) // $ hasValueFlow="xzReader"
+	tarRead = tar.NewReader(xzReader)
+	fmt.Println(io.SeekStart)
+
+	TarDecompressor(tarRead)
+	TarDecompressor2(tarRead)
+	TarDecompressorSafe(tarRead)
+	TarDecompressorTN(tarRead)
+}
+
+func XzSafe(file io.Reader) {
+	xzReader, _ := xz.NewReader(file)
+	var out []byte = make([]byte, 70)
+	var totalBytes int64 = 0
+	i := 1
+	for i > 0 {
+		i, _ = xzReader.Read(out) // GOOD: The output size is being controlled
+		fmt.Println("do sth with output:", xzReader)
+		totalBytes = totalBytes + int64(i)
+		if totalBytes > 1024*1024*1024*5 {
+			fmt.Print(totalBytes, "exceeded")
+			break
+		}
+	}
+}
+
+func TarDecompressor(tarRead *tar.Reader) {
+	files := make(fstest.MapFS)
+	for {
+		cur, err := tarRead.Next()
+		if err == io.EOF {
+			break
+		}
+		if cur.Typeflag != tar.TypeReg {
+			continue
+		}
+		data, _ := io.ReadAll(tarRead) // $ hasValueFlow="tarRead"
+		files[cur.Name] = &fstest.MapFile{Data: data}
+	}
+	fmt.Print(files)
+}
+
+func TarDecompressor2(tarRead *tar.Reader) {
+	var tarOut []byte = make([]byte, 70)
+	tarRead.Read(tarOut) // $ hasValueFlow="tarRead"
+	fmt.Println("do sth with output:", tarOut)
+}
+
+func TarDecompressorTN(tarRead *tar.Reader) {
+	var tarOut []byte = make([]byte, 70)
+	i := 1
+	for i > 0 {
+		i, _ = tarRead.Read(tarOut) // $ hasValueFlow="tarRead"
```

Comment:
Have I understood correctly that this is a TN? If so, you should remove the inline test expectation here, and note that the sanitizer is having an effect.
```suggestion
		i, _ = tarRead.Read(tarOut) // GOOD: the output size is being controlled
```